### PR TITLE
[WIP] Origin refactor

### PR DESCRIPTION
**Pull request by csarofeen**
Uncertain how much this will help, need to test after we merge in https://github.com/csarofeen/pytorch/pull/347 and https://github.com/csarofeen/pytorch/pull/348

It was returning an extra 2us, although it's hard to test without those 2 PRs.

### CODE REVIEW

**Review by naoyam — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/fusion.h [hunk 1]
```diff
@@ -165,7 +165,10 @@ class TORCH_CUDA_API Fusion final {
   std::unordered_set<Expr*> unordered_uses(Val* val) const;
 
   // Return the Expr that produces val
-  Expr* origin(const Val* val) const;
+  Expr* origin(Val* val) const;
+
+  // Return the Expr that produces val
+  const Expr* origin(const Val* val) const;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/fusion.h:171]: 1.  we could have a single overload: `Expr* origin(const Val* val) const`
 (we don't have deep constness in Fusion, so this is perfectly fine)

2. if we go down this route, please add a todo note to completely remove Fusion::origin()

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h [hunk 2]
```diff
@@ -215,8 +215,13 @@ class TORCH_CUDA_API Val : public Statement {
 
   // Returns the Expr that this value is an output of, returns nullptr if none
   // was found
-  Expr* getOrigin();
-  const Expr* getOrigin() const;
+  Expr* getOrigin() {
+    return origin;
+  };
+
+  const Expr* getOrigin() const {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.h:222]: a single overload should be enough (`Expr* origin() const`)

**Review by tlemo — APPROVED** [review]
This seems like a good cleanup regardless of the perf gains.

It also seems a great opportunity to rename 'origin' to 'def' (or definition) to make it consistent with the common compiler terminology.

**Comment by jjsjann123** [comment]
Just to put the comment here that, I tried these change but did not observe any noticeable speed up in our latency.

**Comment by csarofeen** [comment]
Stale, closing.

**Pull request closed without merging**